internal/controller/supervisorconfig/oidcclientwatcher/oidc_client_watcher.go

```diff
@@ -0,0 +1,317 @@
+	"k8s.io/apimachinery/pkg/labels"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+
+	"go.pinniped.dev/generated/latest/apis/supervisor/config/v1alpha1"
+	pinnipedclientset "go.pinniped.dev/generated/latest/client/supervisor/clientset/versioned"
+	configInformers "go.pinniped.dev/generated/latest/client/supervisor/informers/externalversions/config/v1alpha1"
+	pinnipedcontroller "go.pinniped.dev/internal/controller"
+	"go.pinniped.dev/internal/controller/conditionsutil"
+	"go.pinniped.dev/internal/controllerlib"
+	"go.pinniped.dev/internal/oidcclientsecretstorage"
+	"go.pinniped.dev/internal/plog"
+)
+
+const (
+	clientSecretExists     = "ClientSecretExists"
+	allowedGrantTypesValid = "AllowedGrantTypesValid"
+	allowedScopesValid     = "AllowedScopesValid"
+
+	reasonSuccess              = "Success"
+	reasonMissingRequiredValue = "MissingRequiredValue"
+	reasonNoClientSecretFound  = "NoClientSecretFound"
+
+	authorizationCodeGrantTypeName = "authorization_code"
+	refreshTokenGrantTypeName      = "refresh_token"
+	tokenExchangeGrantTypeName     = "urn:ietf:params:oauth:grant-type:token-exchange" //nolint:gosec // this is not a credential
+
+	openidScopeName          = "openid"
+	offlineAccessScopeName   = "offline_access"
+	requestAudienceScopeName = "pinniped:request-audience"
+	usernameScopeName        = "username"
+	groupsScopeName          = "groups"
+
+	allowedGrantTypesFieldName = "allowedGrantTypes"
+	allowedScopesFieldName     = "allowedScopes"
+
+	secretTypeToObserve = "storage.pinniped.dev/oidc-client-secret" //nolint:gosec // this is not a credential
+)
+
+type oidcClientWatcherController struct {
+	pinnipedClient     pinnipedclientset.Interface
+	oidcClientInformer configInformers.OIDCClientInformer
+	secretInformer     corev1informers.SecretInformer
+}
+
+// NewOIDCClientWatcherController returns a controllerlib.Controller that watches OIDCClients and updates
+// their status with validation errors.
+func NewOIDCClientWatcherController(
+	pinnipedClient pinnipedclientset.Interface,
+	secretInformer corev1informers.SecretInformer,
+	oidcClientInformer configInformers.OIDCClientInformer,
+	withInformer pinnipedcontroller.WithInformerOptionFunc,
+) controllerlib.Controller {
+	return controllerlib.New(
+		controllerlib.Config{
+			Name: "OIDCClientWatcherController",
+			Syncer: &oidcClientWatcherController{
+				pinnipedClient:     pinnipedClient,
+				secretInformer:     secretInformer,
+				oidcClientInformer: oidcClientInformer,
+			},
+		},
+		// We want to be notified when an OIDCClient's corresponding secret gets updated or deleted.
+		withInformer(
+			secretInformer,
+			pinnipedcontroller.MatchAnySecretOfTypeFilter(secretTypeToObserve, pinnipedcontroller.SingletonQueue()),
+			controllerlib.InformerOption{},
+		),
+		// We want to be notified when anything happens to an OIDCClient.
+		withInformer(
+			oidcClientInformer,
+			pinnipedcontroller.MatchAnythingFilter(pinnipedcontroller.SingletonQueue()),
```

Comment:
This should filter out anything that does not have the `client.oauth.pinniped.dev-` prefix so that way if we add other clients here in the future this "old" controller does not attempt to handle them.

Reply:
Updated in #1221.